### ISSUE TITLE
How to send chunk data by using vegeta?

### ISSUE THREAD

**Issue by contactparthshah** [tsenart/vegeta#439]
Does vegeta sends chunk data when Transfer-Encoding header is set to "chunked" while sending data?

**Comment by tsenart** [comment]
I don't understand this question.

**Comment by contactparthshah** [comment]
Sorry for that. let me refresh it.

**Comment by tsenart** [comment]
Sorry, refresh what? I know what Chunked transfer encoding is. Can you rephrase your question?

**Comment by contactparthshah** [comment]
Hi tsenart,

can vegeta sends chunked data? Requirement is, server should be able to handle chunked requests  so How can I achieve by using vegeta?

**Comment by tsenart** [comment]
Vegeta sends chunked requests by default when using HTTP 1.1. Are you seeing something different?

**Comment by contactparthshah** [comment]
I haven't tried it yet. Before try out, was gathering an information.

Will behavior different when Content-Length header is set?

**Comment by tsenart** [comment]
Please try it.

**Comment by contactparthshah** [comment]
I tried it out and it seems vegeta sends Content-Length Header

1) I used the below command to send data to the server

~$vegeta attack -lazy=true -timeout=5s -insecure=true -header='Host: 1.1.1.1' -header='Transfer-Encoding: chunked' -targets=targets.txt -format=http

2) On Server side, tcpdump captures shows the headers received are:

POST /ABC/XYZ/1.ts HTTP/1.1
Host: 1.1.1.1
User-Agent: Go-http-client/1.1
Content-Length: 640000
Accept-Encoding: gzip

3) Contents of targets.txt

~$ cat targets.txt
POST http://1.1.1.1/ABC/XYZ/1.ts
@./segment1.ts

~$ ls -l segment1.ts
-rw-r--r-- 1 root root 640000 Oct 14 23:11 segment1.ts
~$

Could you please suggest if something is missing for sending chunk data?

**Comment by contactparthshah** [comment]
@tsenart , can you guide here, what am I missing here? I feel, its an issue

**Comment by tsenart** [comment]
Thanks for reporting this. I added the `-chunked` flag to `vegeta attack`.

**Comment by contactparthshah** [comment]
@tsenart , Currently, I don't see latest vegeta precompiled binary which has this flag. How can I achieve it? Are you planning to create new release version?

**Comment by contactparthshah** [comment]
@tsenart ,

I compiled the binary and tried to verify the fix and not seeing desired behavior:

when -chunked is provided, vegeta is sending "Transfer-Encoding: chunked" header but at the same time actual request body doesn't comes in chunks.

**Comment by tsenart** [comment]
There's a debug server in the vegeta repo you can use to tell me what vegeta is sending.

In one terminal window / tab.
```
cd vegeta
go run ./internal/cmd/echosrv/ -dump :6060
```

In another terminal window / tab.
```
echo "HELLO WORLD" > body.txt
echo "POST http://127.0.0.1:6060/" | ./vegeta attack -chunked -rate=1 -body=body.txt  | ./vegeta report --every 1s
```

The server dumps the raw incoming request to stdout. What do you see? Isn't the body chunked?

**Comment by contactparthshah** [comment]
@tsenart , whole data comes as one chunk only.

with chunk
`POST / HTTP/1.1
Host: 127.0.0.1:6060
Transfer-Encoding: chunked
Accept-Encoding: gzip
User-Agent: Go-http-client/1.1

c
HELLO WORLD

0
`

without chunk
`
POST / HTTP/1.1
Host: 127.0.0.1:6060
Accept-Encoding: gzip
Content-Length: 12
User-Agent: Go-http-client/1.1

HELLO WORLD
`
Is there anyway to specify divide request body into few chunks like divide request body into 2 or 3 chunks.

If we take reference of example [here](https://en.wikipedia.org/wiki/Chunked_transfer_encoding) - then request body is divided into 3 chunks.
`
4\r\n
Wiki\r\n
5\r\n
pedia\r\n
E\r\n
 in\r\n
\r\n
chunks.\r\n
0\r\n
\r\n
`

**Comment by tsenart** [comment]
What makes it necessary for the chunks to be a certain size or certain number?

**Comment by contactparthshah** [comment]
Server resources become important how it behaves When client sends data in chunk. Currently, data is definitely coming in chunks (--chunked) but there is no real difference with and without chunk as Whole data is coming in one chunk only. Example: Let's say when data is divided into 3 chunks so it might be possible that 2 chunks come fast and 3rd chunk come after some time so during that Server connection left open and when multiple of these requests are there it becomes important how server behaves. I believe, it will be good if we can give some option on certain number for chunk.

**Comment by tsenart** [comment]
Understood. Re-opening this issue to address this problem.

**Comment by contactparthshah** [comment]
@tsenart , let me know if you need help for doing verification.
